### PR TITLE
Add a function to update FLAGS

### PR DESCRIPTION
**Pull request by gfwm2013**
目前FLAGS的设置需在import paddle之前通过设置environment的方式进行，如模型库中PaddleCV/ssd模型[train.py](https://github.com/PaddlePaddle/models/blob/1dfba509fd3880f16b949f5584b2eac1cb159496/PaddleCV/ssd/train.py#L24)的代码一样
image

现在可以通过`set_flags`函数设置flags的值，无需要求在import paddle之前使用，具体用法：
```
flags = {'FLAGS_eager_delete_tensor_gb': 1.0}
fluid.set_flags(flags)
```
并且可以通过`get_flags`接口来获取相关flag的值，具体用法：
```
flags = ['FLAGS_eager_delete_tensor_gb']
res = fluid.get_flags(flags)
print(res)   # {'FLAGS_eager_delete_tensor_gb': 1.0}
```

172 18 25 111_8088_documentation_docs_en_api_fluid_get_flags html (1)
172 18 25 111_8088_documentation_docs_en_api_fluid_get_flags html

### CODE REVIEW

File: paddle/fluid/pybind/global_value_getter_setter.cc [hunk 1]
```diff
@@ -32,12 +32,44 @@ DECLARE_bool(use_ngraph);
...
+DECLARE_double(memory_fraction_of_eager_deletion);
+DECLARE_bool(use_pinned_memory);
+DECLARE_bool(benchmark);
+DECLARE_int32(inner_op_parallelism);
+DECLARE_string(tracer_profile_fname);
+DECLARE_string(allocator_strategy);
 #ifdef PADDLE_WITH_CUDA
+// DECLARE_int32(rpc_send_thread_num);
+// DECLARE_int32(rpc_get_thread_num);
+// DECLARE_int32(rpc_prefetch_thread_num);
```
Comment (zhhsplendid) [inline paddle/fluid/pybind/global_value_getter_setter.cc]: Why do you comment some "DECLARE" flags?

File: paddle/fluid/pybind/global_value_getter_setter.cc [hunk 2]
```diff
@@ -53,21 +85,34 @@ class PYBIND11_HIDDEN GlobalVarGetterSetterRegistry {
   using Setter = std::function<void(const py::object &)>;
   using Getter = std::function<py::object()>;
 
+  struct Flag_Getter {
+    bool flag;
```
Comment (zhhsplendid) [inline paddle/fluid/pybind/global_value_getter_setter.cc]: You had better rename or add comment about what's the bool "flag".

For example, a code reader will be confused about what's bool "flag" means? Jinle and I understand "flag" means "is_public" but that may not be true for other SWEs

File: paddle/fluid/pybind/global_value_getter_setter.cc [hunk 3]
```diff
@@ -53,21 +85,34 @@ class PYBIND11_HIDDEN GlobalVarGetterSetterRegistry {
   using Setter = std::function<void(const py::object &)>;
   using Getter = std::function<py::object()>;
 
+  struct Flag_Getter {
```
Comment (zhhsplendid) [inline paddle/fluid/pybind/global_value_getter_setter.cc]: Why "Flag_Getter"? Isn't "FlagGetter" a proper name?

File: paddle/fluid/pybind/global_value_getter_setter.cc [hunk 4]
```diff
@@ -144,44 +240,104 @@ void BindGlobalValueGetterSetter(pybind11::module *module) {
 
   py::class_<GlobalVarGetterSetterRegistry>(*module,
                                             "GlobalVarGetterSetterRegistry")
-      .def("__getitem__", &GlobalVarGetterSetterRegistry::Get)
-      .def("__setitem__", &GlobalVarGetterSetterRegistry::Set)
+      .def("__getpublic__", &GlobalVarGetterSetterRegistry::GetPublic)
```
Comment (zhhsplendid) [inline paddle/fluid/pybind/global_value_getter_setter.cc]: Why do you make `__XXX__` for these methods? Do you know these kind of naming is usually used for special python defined function/variables?

For example, if you search `__contains__, __setitem__, __getitem__`, you will know they are  python defined methods

File: paddle/fluid/pybind/global_value_getter_setter.cc [hunk 5]
```diff
@@ -144,44 +240,104 @@ void BindGlobalValueGetterSetter(pybind11::module *module) {
...
+      .def("getprivate",
+           &GlobalVarGetterSetterRegistry::GetPrivateOrReturnDefaultValue,
            py::arg("key"), py::arg("default") = py::cast<py::none>(Py_None));
 
   module->def("globals", &GlobalVarGetterSetterRegistry::Instance,
               py::return_value_policy::reference);
 }
 
-#define REGISTER_GLOBAL_VAR_GETTER_ONLY(var)                        \
+#define REGISTER_PRIVATE_GLOBAL_VAR_GETTER_ONLY(var, flag)          \
```
Comment (zhhsplendid) [inline paddle/fluid/pybind/global_value_getter_setter.cc]: In fact, your "var" is a flag and your 'flag' means 'is_public' ...

Rename 'flag's here around

File: python/paddle/fluid/framework.py [hunk 6]
```diff
@@ -5136,3 +5138,58 @@ def device_guard(device=None):
...
     yield
     switch_device(pre_device)
+
+
+def set_flags(flags):
+    """
+    This function sets the GFlags value in Paddle.
+
+    Args:
+        flags (dict): A dict contains flags and it's value.
```
Comment (zhhsplendid) [inline python/paddle/fluid/framework.py]: 'it's' -> 'its'

File: python/paddle/fluid/framework.py [hunk 7]
```diff
@@ -5136,3 +5138,58 @@ def device_guard(device=None):
...
+
+    Examples:
+            .. code-block:: python
+
+                import paddle.fluid as fluid
+                fluid.set_flags({'FLAGS_FLAGS_eager_delete_tensor_gb': 1.0})
```
Comment (zhhsplendid) [inline python/paddle/fluid/framework.py]: FLAGS_FLAGS_eager_delete_tensor_gb ? Is it a typo? Double "FLAGS" at the beginning?

File: paddle/fluid/pybind/global_value_getter_setter.cc (hunk 3 above)
Comment (gfwm2013) [reply paddle/fluid/pybind/global_value_getter_setter.cc]: Done.

File: paddle/fluid/pybind/global_value_getter_setter.cc (hunk 5 above)
Comment (gfwm2013) [reply paddle/fluid/pybind/global_value_getter_setter.cc]: Done.

File: python/paddle/fluid/framework.py (hunk 6 above)
Comment (gfwm2013) [reply python/paddle/fluid/framework.py]: Done.

File: python/paddle/fluid/framework.py (hunk 7 above)
Comment (gfwm2013) [reply python/paddle/fluid/framework.py]: Done.

File: paddle/fluid/pybind/global_value_getter_setter.cc (hunk 1 above)
Comment (zhhsplendid) [reply paddle/fluid/pybind/global_value_getter_setter.cc]: After discussion, it is DISTRIBUTION flags, please add WITH_DISTRIBUTION macros.

File: paddle/fluid/pybind/global_value_getter_setter.cc [hunk 8]
```diff
@@ -32,12 +32,44 @@ DECLARE_bool(use_ngraph);
...
 DECLARE_bool(cudnn_deterministic);
+DECLARE_uint64(conv_workspace_size_limit);
+DECLARE_bool(cudnn_batchnorm_spatial_persistent);
+DECLARE_bool(cudnn_exhaustive_search);
+DECLARE_bool(eager_delete_scope);
+DECLARE_bool(fast_eager_deletion_mode);
+DECLARE_double(fraction_of_cuda_pinned_memory_to_use);
+DECLARE_double(fraction_of_gpu_memory_to_use);
+DECLARE_uint64(initial_gpu_memory_in_mb);
+// DECLARE_int64(reallocate_gpu_memory_in_mb);
```
Comment (zhhsplendid) [inline paddle/fluid/pybind/global_value_getter_setter.cc]: DECLARE_int64(reallocate_gpu_memory_in_mb) -> DECLARE_uint64(reallocate_gpu_memory_in_mb)

File: paddle/fluid/pybind/global_value_getter_setter.cc [hunk 9]
```diff
@@ -144,44 +242,102 @@ void BindGlobalValueGetterSetter(pybind11::module *module) {
...
+  REGISTER_PUBLIC_GLOBAL_VAR_GETTER_SETTER(FLAGS_eager_delete_scope, true);
+  REGISTER_PUBLIC_GLOBAL_VAR_GETTER_SETTER(FLAGS_fast_eager_deletion_mode,
+                                           true);
+  REGISTER_PUBLIC_GLOBAL_VAR_GETTER_SETTER(
+      FLAGS_fraction_of_cuda_pinned_memory_to_use, true);
+  REGISTER_PUBLIC_GLOBAL_VAR_GETTER_SETTER(FLAGS_fraction_of_gpu_memory_to_use,
+                                           true);
+  REGISTER_PUBLIC_GLOBAL_VAR_GETTER_SETTER(FLAGS_initial_gpu_memory_in_mb,
+                                           true);
+  // REGISTER_PUBLIC_GLOBAL_VAR_GETTER_SETTER(FLAGS_reallocate_gpu_memory_in_mb,
```
Comment (zhhsplendid) [inline paddle/fluid/pybind/global_value_getter_setter.cc]: Change these flags as I commented at the head of this file

File: paddle/fluid/pybind/global_value_getter_setter.cc [hunk 10]
```diff
@@ -33,12 +33,46 @@ DECLARE_bool(use_ngraph);
 DECLARE_bool(use_system_allocator);
 DECLARE_bool(free_idle_chunk);
 DECLARE_bool(free_when_no_cache_hit);
+DECLARE_bool(check_nan_inf);
```
Comment (zhhsplendid) [inline paddle/fluid/pybind/global_value_getter_setter.cc:33]: It is fine now, but in the future you may group flags together.

Usually, flags/imports are grouped by their functionality first, then alphabetical order

File: python/paddle/fluid/tests/unittests/test_get_set_flags.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        self.assertTrue(res['FLAGS_eager_delete_tensor_gb'], 1.0)
+
+
+class TestGetAndSetFlagsErrors(unittest.TestCase):
+    def test_errors(self):
+        flags_list = ['FLAGS_eager_delete_tensor_gb', 'FLAGS_check_nan_inf']
+        flag = 1
+        flag_private = {'FLAGS_use_mkldnn': True}
+
+        # flags type is set_flags should be dict.
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_get_set_flags.py]: 'is' -> 'of'

File: python/paddle/fluid/tests/unittests/test_get_set_flags.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        self.assertRaises(TypeError, test_set_flags_input_type)
+
+        # flags in set_flags should be public flags.
+        def test_set_private_flag():
+
+            fluid.get_flags('FLAGS_use_mkldnn')
+
+        self.assertRaises(ValueError, test_set_private_flag)
+
+        # flags type is set_flags should be list, tuple or string
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_get_set_flags.py]: 'is' -> 'of'

**Review by zhhsplendid — DISMISSED** [review]
There is tiny, not important issues in this PR. I will approve it and you can change in the future or later PR when you have time

**Review by sneaxiy — DISMISSED** [review]
LGTM.

File: paddle/fluid/pybind/global_value_getter_setter.cc [hunk 13]
```diff
@@ -290,13 +324,32 @@ static void RegisterGlobalVarGetterSetter() {
...
+      FLAGS_fast_eager_deletion_mode,
+      FLAGS_fraction_of_cuda_pinned_memory_to_use,
+      FLAGS_fraction_of_gpu_memory_to_use, FLAGS_initial_gpu_memory_in_mb,
+      FLAGS_reallocate_gpu_memory_in_mb, FLAGS_enable_cublas_tensor_op_math,
+      FLAGS_selected_gpus, FLAGS_sync_nccl_allreduce);
+#endif
+#ifdef PADDLE_WITH_DITRIBUTE
+  REGISTER_PUBLIC_GLOBAL_VAR(FLAGS_rpc_send_thread_num,
+                             FLAGS_rpc_get_thread_num,
+                             FLAGS_rpc_prefetch_thread_num)
```
Comment (sneaxiy) [inline paddle/fluid/pybind/global_value_getter_setter.cc]: Add `;` ?

File: paddle/fluid/pybind/global_value_getter_setter.cc (hunk 1 above)
Comment (gfwm2013) [reply paddle/fluid/pybind/global_value_getter_setter.cc]: Done.

File: paddle/fluid/pybind/global_value_getter_setter.cc (hunk 8 above)
Comment (gfwm2013) [reply paddle/fluid/pybind/global_value_getter_setter.cc]: Done.

File: paddle/fluid/pybind/global_value_getter_setter.cc (hunk 9 above)
Comment (gfwm2013) [reply paddle/fluid/pybind/global_value_getter_setter.cc]: Done.

File: paddle/fluid/pybind/global_value_getter_setter.cc (hunk 10 above)
Comment (gfwm2013) [reply paddle/fluid/pybind/global_value_getter_setter.cc:33]: Done.

File: paddle/fluid/pybind/global_value_getter_setter.cc (hunk 13 above)
Comment (gfwm2013) [reply paddle/fluid/pybind/global_value_getter_setter.cc]: Done.

File: python/paddle/fluid/tests/unittests/test_get_set_flags.py (hunk 11 above)
Comment (gfwm2013) [reply python/paddle/fluid/tests/unittests/test_get_set_flags.py]: Done.

File: python/paddle/fluid/tests/unittests/test_get_set_flags.py (hunk 12 above)
Comment (gfwm2013) [reply python/paddle/fluid/tests/unittests/test_get_set_flags.py]: Done.

**Review by lanxianghit — APPROVED** [review]

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Comment by zhhsplendid** [comment]
close and re-open to trigger CI

**Review by Boyan-Liu — APPROVED** [review]

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Pull request merged**